# 1.


In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
df = pd.read_csv(url)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [4]:
df.isna().sum()

#               0
Name            0
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

In [5]:
df.isna()

#   Name  Type 1  Type 2     HP  Attack  Defense  Sp. Atk  Sp. Def  \
0    False  False   False   False  False   False    False    False    False   
1    False  False   False   False  False   False    False    False    False   
2    False  False   False   False  False   False    False    False    False   
3    False  False   False   False  False   False    False    False    False   
4    False  False   False    True  False   False    False    False    False   
..     ...    ...     ...     ...    ...     ...      ...      ...      ...   
795  False  False   False   False  False   False    False    False    False   
796  False  False   False   False  False   False    False    False    False   
797  False  False   False   False  False   False    False    False    False   
798  False  False   False   False  False   False    False    False    False   
799  False  False   False   False  False   False    False    False    False   

     Speed  Generation  Legendary  
0    False       False      False  
1    False       False      False  
2    False       False      False  
3    False       False      False  
4    False       False      False  
..     ...         ...        ...  
795  False       False      False  
796  False       False      False  
797  False       False      False  
798  False       False      False  
799  False       False      False  

[800 rows x 12 columns]

# 2.

In [8]:
size1 = df.shape
print("The size from the dataframe is: ", size1)

The size from the dataframe is:  (800, 12)


In [11]:
col = df.columns.tolist()
row = len(df)
print("Columns are: ", col)
print("rows are: ", row)

Columns are:  ['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary']
rows are:  800


In [13]:
df.describe()

#          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.000000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750   69.258750   79.001250   73.842500   72.820000   71.902500   
std    208.343798   25.534669   32.457366   31.183501   32.722294   27.828916   
min      1.000000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    184.750000   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    364.500000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    539.250000   80.000000  100.000000   90.000000   95.000000   90.000000   
max    721.000000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000000     6.00000

# 3.

In [15]:
df['Type 1'].value_counts()

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

In [22]:
df.describe()

#          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.000000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750   69.258750   79.001250   73.842500   72.820000   71.902500   
std    208.343798   25.534669   32.457366   31.183501   32.722294   27.828916   
min      1.000000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    184.750000   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    364.500000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    539.250000   80.000000  100.000000   90.000000   95.000000   90.000000   
max    721.000000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000000     6.00000

# 4.

In [6]:
df.shape

(800, 12)

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   HP          800 non-null    int64 
 5   Attack      800 non-null    int64 
 6   Defense     800 non-null    int64 
 7   Sp. Atk     800 non-null    int64 
 8   Sp. Def     800 non-null    int64 
 9   Speed       800 non-null    int64 
 10  Generation  800 non-null    int64 
 11  Legendary   800 non-null    bool  
dtypes: bool(1), int64(8), object(3)
memory usage: 69.7+ KB


In [16]:
df['Type 2'].value_counts()

Type 2
Flying      97
Ground      35
Poison      34
Psychic     33
Fighting    26
Grass       25
Fairy       23
Steel       22
Dark        20
Dragon      18
Water       14
Ghost       14
Ice         14
Rock        14
Fire        12
Electric     6
Normal       4
Bug          3
Name: count, dtype: int64

In [25]:
import numpy as np

In [26]:
df.shape[0]

800


When working with datasets that contain both non-numeric variables and missing values in numeric variables, you may notice discrepancies between the size of the dataset reported by `df.shape` and the statistics reported by `df.describe()`. Let’s break down these discrepancies:

### 1. Number of Columns Analyzed

- **`df.shape`**: This provides the overall dimensions of the DataFrame, including the number of rows and columns.
  
  ```python
  print(df.shape)
  ```

  This will output a tuple, e.g., `(800, 12)`, meaning there are 800 rows and 12 columns.

- **`df.describe()`**: This method provides summary statistics, but only for numerical columns by default. It does not include non-numeric columns such as strings or categorical variables.

  ```python
  print(df.describe())
  ```

  If your DataFrame has 12 columns but only 8 of them are numeric, `df.describe()` will show statistics for those 8 numeric columns only.

### 2. Values in the "Count" Column

- **In `df.shape`**: The total number of rows (reported as part of the shape) represents all rows in the DataFrame, regardless of whether there are missing values.

  ```python
  print(df.shape[0])  # Number of rows
  ```

  This will give you the total number of rows, e.g., 800.

- **In `df.describe()`**: The "count" column in the output of `df.describe()` represents the number of non-null (non-missing) values for each numeric column. If a numeric column has missing values, the count will be less than the total number of rows.

  ```python
  print(df.describe())
  ```

  For example, if you have a column with 800 total rows but 50 missing values, the "count" for that column will be 750, not 800.

### Summary

- **Number of Columns**: `df.shape` includes all columns, whereas `df.describe()` only includes numeric columns by default.

- **Count Values**: `df.shape` shows the total number of rows, while `df.describe()` shows the count of non-null values for each numeric column, which may be less than the total number of rows if there are missing values.

If you need to analyze non-numeric columns or understand missing values better, you can use methods like `df.info()` for overall structure and `df.isna().sum()` to count missing values in all columns.


# 5.

##
For attrbutes: These are connected within a variable and an object. 
               They are representing the data or values that stored within an object.
               Likewise, df.shape gives us a tuple to represent the size from a dataframe.

##
For methods: These are f(x) to an object. They are used to compute and actions.
             df.head() computes a fuction and return the first few rows of the dataframe.

# 6.

The `df.describe()` method in `pandas` provides a set of summary statistics for each numeric variable it analyzes. Here's a definition of each statistic:

### 1. **Count**:
   - **Definition**: The number of non-null (non-missing) values in the column. This does not include any `NaN` values.
   - **Interpretation**: It indicates how many valid data points are present for a particular variable.

### 2. **Mean**:
   - **Definition**: The average value of the column, calculated by summing all the non-null values and dividing by the count of those values.
   - **Formula**: 
     \[
     \text{Mean} = \frac{\sum_{i=1}^{n} x_i}{n}
     \]
     where \(x_i\) are the values and \(n\) is the count.
   - **Interpretation**: It gives a central measure of the data, indicating where the "middle" of the data lies.

### 3. **Standard Deviation (std)**:
   - **Definition**: A measure of the amount of variation or dispersion in a dataset. It quantifies how spread out the values are from the mean.
   - **Formula**: 
     \[
     \text{Std} = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (x_i - \text{mean})^2}
     \]
   - **Interpretation**: A smaller standard deviation means the data points are closer to the mean, while a larger standard deviation means they are more spread out.

### 4. **Minimum (min)**:
   - **Definition**: The smallest non-null value in the column.
   - **Interpretation**: It represents the lowest observed value in the dataset for that particular variable.

### 5. **25th Percentile (25%)**:
   - **Definition**: The value below which 25% of the data points fall. Also known as the **first quartile (Q1)**.
   - **Interpretation**: It gives a sense of where the lower quartile of the data lies, helping to understand the spread of the lower half of the data.

### 6. **50th Percentile (50%) / Median**:
   - **Definition**: The middle value of the data, where 50% of the data points are below and 50% are above. It is also known as the **median**.
   - **Interpretation**: The median is a robust measure of central tendency and is less affected by extreme values than the mean.

### 7. **75th Percentile (75%)**:
   - **Definition**: The value below which 75% of the data points fall. Also known as the **third quartile (Q3)**.
   - **Interpretation**: It helps in understanding the upper quartile of the data, showing where the bulk of the higher values are distributed.

### 8. **Maximum (max)**:
   - **Definition**: The largest non-null value in the column.
   - **Interpretation**: It represents the highest observed value in the dataset for that particular variable.

### Example Interpretation:

If you had a column of data representing the weight of Pokémon, the summary statistics might look like this:

|       | weight (kg) |
|-------|-------------|
| count | 800         |
| mean  | 50          |
| std   | 25          |
| min   | 1           |
| 25%   | 30          |
| 50%   | 50          |
| 75%   | 70          |
| max   | 120         |

- **Count**: 800 non-null weights.
- **Mean**: The average weight is 50 kg.
- **Std**: The standard deviation is 25, indicating some spread in the weights.
- **Min**: The lightest Pokémon weighs 1 kg.
- **25%**: 25% of Pokémon weigh less than 30 kg.
- **50%**: Half of the Pokémon weigh less than 50 kg.
- **75%**: 75% of Pokémon weigh less than 70 kg.
- **Max**: The heaviest Pokémon weighs 120 kg.

These statistics give you a concise summary of the distribution and spread of your data.

# 7.

#### If you use del df['col'], it would delete an entire column, regardless of whether only a few values are missing or the entire column is incomplete.

## ### Opposite Use Case: When Using `del df['col']` is Preferred Over `df.dropna()`

#### Scenario:
Imagine you are working with a dataset that contains a column where almost all values are missing or irrelevant for your analysis. Instead of removing rows that have missing values in this column (which would reduce the dataset size significantly), you decide to completely remove the column itself because it's not useful.

### Preferred Method: `del df['col']`

```python
# Delete the 'Speed' column, which has too many missing values
del df['Speed']
```

- **Why It's Preferred**: If the column has too many missing values or is irrelevant to the analysis, removing the entire column is more efficient than removing rows, which could result in losing valuable data from other columns.
  
#### Example:
Let’s consider the following DataFrame:

| Name       | Attack | Defense | Speed |
|------------|--------|---------|-------|
| Pikachu    | 55     | 40      | 90    |
| Bulbasaur  | 49     | 49      | NaN   |
| Charmander | 52     | 43      | NaN   |
| Squirtle   | 48     | 65      | NaN   |

In this example, the "Speed" column has many missing values. Rather than dropping rows, you decide to delete the entire column because it's not needed for the analysis.

### Why `del df['col']` Before `df.dropna()` Could Be Important

If you have columns with a significant number of missing values, deleting the column using `del df['col']` before applying `df.dropna()` is important because:
1. **Preserving Rows**: If you apply `df.dropna()` without first deleting the column with many missing values, it will remove many rows, reducing the dataset size and potentially discarding useful data in other columns.
2. **Efficient Cleanup**: By deleting irrelevant columns first, you ensure that the `dropna()` operation focuses on rows with meaningful data rather than rows that are incomplete due to unimportant columns.

### Example: Removing All Missing Data From a Dataset

Now let’s apply this logic to the Pokémon dataset.

#### Step 1: Load the Dataset and Check for Missing Values

```python
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv'
df = pd.read_csv(url)

# Check for missing values before applying any operations
print("Missing Values Before Cleanup:")
print(df.isna().sum())
```

#### Step 2: Use `del df['col']` to Remove Columns with Many Missing Values

Let’s assume that some columns (e.g., `Type 2`) have too many missing values and are not necessary for your analysis.

```python
# Delete the 'Type 2' column because it has many missing values
del df['Type 2']

# Check missing values after deleting the column
print("\nMissing Values After Removing 'Type 2' Column:")
print(df.isna().sum())
```

#### Step 3: Apply `df.dropna()` to Remove Remaining Rows with Missing Values

Now that we've deleted the columns with excessive missing values, we can safely remove rows that still contain missing data.

```python
# Drop rows with any remaining missing values
df_cleaned = df.dropna()

# Check missing values after dropping rows
print("\nMissing Values After Dropping Rows with Missing Data:")
print(df_cleaned.isna().sum())
```

### Before and After Report

#### Before:
After loading the dataset and checking for missing values, the report might look like this:

```
Missing Values Before Cleanup:
# Example output for demonstration (actual dataset may vary)
Type 2       200
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       0
Speed         0
```

In this case, the "Type 2" column has 200 missing values, but the other columns have no missing values.

#### After Deleting Column and Dropping Rows:
After cleaning the dataset using a combination of `del df['Type 2']` and `df.dropna()`:

```
Missing Values After Removing 'Type 2' Column:
HP          0
Attack      0
Defense     0
Sp. Atk     0
Sp. Def     0
Speed       0

Missing Values After Dropping Rows with Missing Data:
HP          0
Attack      0
Defense     0
Sp. Atk     0
Sp. Def     0
Speed       0
```

### Justification for Approach:

1. **Removing Irrelevant Data**: The "Type 2" column had too many missing values and was not essential to the analysis, so deleting it reduced the amount of noise in the dataset without losing too much information.
   
2. **Preserving Useful Data**: By deleting the column first, I preserved the rows that had complete and useful data in other columns, ensuring that valuable information was not unnecessarily discarded.

3. **Efficient Cleanup**: After deleting the column, `df.dropna()` was applied to remove rows with any remaining missing values, leaving behind a clean dataset ready for analysis.

This combination of steps is efficient for datasets where some columns may be irrelevant or contain too many missing values, and you want to ensure that you’re not losing too much data by only removing the problematic parts.

# 8(1).

### df.groupby("col1")["col2"].describe() is  to generate summary statistics for specific column["col2"], its breaking down by groups which defined in column("col1"). It can surely provides of how variables behaves across different categories.

In [9]:
df.groupby("Type 1")["Type 2"]

In [10]:
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

### For the "Titanic" example, by using the df.groupby("col1")["col2"].describe() f(x) can calculate the stats from the category "age", which grouped individually. It does provides the detailed information from each classes ages, such as average ages.

[References](https://chatgpt.com/c/66e350bd-7248-8002-baa3-902a2326585c)